### PR TITLE
[OCPCLOUD-1042] Increase test coverage in CSR approval

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Porting introduced testst from https://github.com/openshift/cluster-machine-approver/pull/73/commits/a7a33f99fe82374879b54d9fa558628fd9e9ff5d as a part of https://github.com/openshift/cluster-machine-approver/pull/73, which are adding full coverage of `csr_check.go` package.
```
Before
ok  	github.com/openshift/cluster-machine-approver	7.143s	coverage: 46.7% of statements
After
ok  	github.com/openshift/cluster-machine-approver	(cached)	coverage: 63.7% of statements
```

Depends on https://github.com/openshift/cluster-machine-approver/pull/96

### CODE REVIEW

**Comment by Danil-Grigorev** [comment]
/hold 

This PR requires golang-1.15 in CI. Otherwise we get an error message mismatch:
Expected
```x509: certificate has expired or is not yet valid: current time 2020-11-18T00:00:00Z is before 2020-11-18T20:12:00Z```
and CI returns just
```x509: certificate has expired or is not yet valid```

File: csr_check_test.go
Comment (elmiko) [inline csr_check_test.go:1405]: i'm curious why some of these entries have the `nodeErr` removed?

File: csr_check_test.go
Comment (elmiko) [inline csr_check_test.go:2421]: should this be `wrongAddr`? 

i might be slightly confused byt the test name/logic

File: csr_check_test.go
Comment (elmiko) [inline csr_check_test.go:2448]: is there any chance we would deadlock here inside of `respond`? (i saw an `accept` call in that function)

File: csr_check_test.go
Comment (elmiko) [inline csr_check_test.go:2250]: is there any chance we would deadlock here inside of `respond`? (i saw an `accept` call in that function)

**Review by elmiko — COMMENTED** [review]
this is _mostly_ making sense to me Danil, i just have a few questions to make it clearer for me.

File: csr_check_test.go
Comment (Danil-Grigorev) [reply csr_check_test.go:2448]: Accept returns an error, once the socket is closed. I ignore it, as I only needed a consumer to return some response with peer certificates (if it is needed, and I call it only when it is needed)

File: csr_check_test.go
Comment (Danil-Grigorev) [reply csr_check_test.go:2250]: https://github.com/openshift/cluster-machine-approver/pull/95#discussion_r527597436

File: csr_check.go
Comment (Danil-Grigorev) [reply csr_check.go:362]: Right, we need to test the certificate exchange flow, and so I had to stub `time.Now` -https://github.com/openshift/cluster-machine-approver/pull/95/files#diff-3c8388bac66caa11eeb9a56a4cb51f879caff503f20824b7bd973f7969eda6ddR412

File: csr_check_test.go
Comment (Danil-Grigorev) [reply csr_check_test.go:1405]: These tests need a lot of refactoring. These test cases as you can see contain no changes aside from this, and they don't fail. So it was initially a copy-paste job, and it is a dead code.

File: csr_check.go
Comment (JoelSpeed) [inline csr_check.go:76]: This error message could really do with improvement couldn't it 🤔

File: csr_check.go
Comment (JoelSpeed) [reply csr_check.go:363]: Can we use a name that doesn't shadow the top level please?

File: main.go
Comment (JoelSpeed) [inline main.go:164]: What's the motivation for moving this outside of the function? I'm not a huge fan of moving logic around to suit the test suites, there are better ways we can do this, eg

```
type indexLister interface {
  List() []interface{}
}

func recentlyPendingCSRs(csrs indexLister) []*CertificateSigningRequest {
...
}

// In test code

type fakeIndexLister struct {
 items []interface{}
}

func (f fakeIndexList) List() ([]interface{}) {
  return f.items
}
```

File: metrics.go
Comment (JoelSpeed) [inline metrics.go:46]: The fact we had to move this twice, definitely suggests we should be making the tests bend to the code rather than the other way around

File: csr_check_test.go
Comment (JoelSpeed) [inline csr_check_test.go:2523]: Looking at the way these tests are structured, we could easily use a fake lister here to avoid having to move the logic in the main code flow

File: csr_check_test.go
Comment (JoelSpeed) [inline csr_check_test.go:2911]: Am I right in thinking `Accept` will return an error if the listener is shutdown?

File: main.go
Comment (Danil-Grigorev) [reply main.go:164]: The goal of this PR is to transition to https://github.com/openshift/cluster-machine-approver/pull/73 and do not change existing tests. The thing you pointing out makes sense to me, but I didn't think about it as a long-term solution, just wanted to preserve tests.

Comment (JoelSpeed) [reply main.go:164]: I don't think it would take too long to do, and would be the right thing given how long it might take to merge things across to that other PR, but it's not a blocker if that's going away in #73

File: csr_check_test.go
Comment (Danil-Grigorev) [reply csr_check_test.go:2911]: It will. I just needed a consumer, so ignoring all the errors worked for me in such a scenario. One pass is guaranteed to be a success, so the code accepts and responds only when it's needed.

File: csr_check.go
Comment (Danil-Grigorev) [reply csr_check.go:80]: I'll work on this in https://github.com/openshift/cluster-machine-approver/pull/73

**Review by elmiko — APPROVED** [review]
i'm good with this, the only outstanding question i have is did Joel's request for changing the name on `now` get picked up?
/lgtm

**Comment by JoelSpeed** [comment]
/approve

Please make sure that #73 picks up the outstanding changes that I had asked for re not having to move logic out of functions to help tests

**Pull request merged**
